src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -0,0 +1,280 @@
+    {
+        private static readonly ParallelOptions DefaultParallelOptions = new();
+
+        public void Initialize(GeneratorInitializationContext context)
+        {
+        }
+
+        public void Execute(GeneratorExecutionContext context)
+        {
+            var razorContext = RazorSourceGenerationContext.Create(context);
+            if (razorContext is null ||
+                (razorContext.RazorFiles.Count == 0 && razorContext.CshtmlFiles.Count == 0))
+            {
+                return;
+            }
+
+            HandleDebugSwitch(razorContext.WaitForDebugger);
+
+            var tagHelpers = ResolveTagHelperDescriptors(context, razorContext);
+
+            var projectEngine = RazorProjectEngine.Create(razorContext.Configuration, razorContext.FileSystem, b =>
+            {
+                b.Features.Add(new DefaultTypeNameFeature());
+                b.SetRootNamespace(razorContext.RootNamespace);
+
+                b.Features.Add(new StaticTagHelperFeature { TagHelpers = tagHelpers, });
+                b.Features.Add(new DefaultTagHelperDescriptorProvider());
+
+                CompilerFeatures.Register(b);
+                RazorExtensions.Register(b);
+
+                b.SetCSharpLanguageVersion(((CSharpParseOptions)context.ParseOptions).LanguageVersion);
+            });
+
+            CodeGenerateRazorComponents(context, razorContext, projectEngine);
+            GenerateViews(context, razorContext, projectEngine);
+        }
+
+        private void GenerateViews(GeneratorExecutionContext context, RazorSourceGenerationContext razorContext, RazorProjectEngine projectEngine)
+        {
+            var files = razorContext.CshtmlFiles;
+
+            Parallel.For(0, files.Count, GetParallelOptions(), i =>
+            {
+                var file = files[i];
+
+                var codeDocument = projectEngine.Process(projectEngine.FileSystem.GetItem(file.NormalizedPath, FileKinds.Legacy));
+                var csharpDocument = codeDocument.GetCSharpDocument();
+                for (var j = 0; j < csharpDocument.Diagnostics.Count; j++)
+                {
+                    var razorDiagnostic = csharpDocument.Diagnostics[j];
+                    var csharpDiagnostic = razorDiagnostic.AsDiagnostic();
+                    context.ReportDiagnostic(csharpDiagnostic);
+                }
+
+                Directory.CreateDirectory(Path.GetDirectoryName(file.GeneratedOutputPath));
+                File.WriteAllText(file.GeneratedOutputPath, csharpDocument.GeneratedCode);
+            });
+        }
+
+        private static void CodeGenerateRazorComponents(GeneratorExecutionContext context, RazorSourceGenerationContext razorContext, RazorProjectEngine projectEngine)
+        {
+            var files = razorContext.RazorFiles;
+
+            var arraypool = ArrayPool<(string, SourceText)>.Shared;
+            var outputs = arraypool.Rent(files.Count);
+
+            Parallel.For(0, files.Count, GetParallelOptions(), i =>
```

Comment:
Please make sure to pass down `context.CancellationToken` for parallelization work like this.

Reply:
Addressed in https://github.com/dotnet/sdk/pull/15909.